plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java

```diff
@@ -89,7 +120,20 @@ public void backupWorkspace(final String workspaceId,
     public void backupWorkspaceAndCleanup(final String workspaceId,
                                           final String srcPath,
                                           final String srcAddress) throws ServerException {
-        backupWorkspace(workspaceId, srcPath, srcAddress, true);
+        ReentrantLock lock = workspacesBackupLocks.get(workspaceId);
+        if (lock != null) {
+            lock.lock();
+            if (workspacesBackupLocks.get(workspaceId) == null) {
+                // should never happen
```

Comment:
Please describe in comment possible situation when it is possible as in https://github.com/codenvy/codenvy/pull/425/files#diff-6fa2b3319d5dc836fd809bbba1ed5605R95